syntax = "proto3";
package pos_rpc;

service Health {
    rpc Check(HealthCheckRequest) returns (HealthCheckResponse);
}

message HealthCheckRequest {
    // nothing
}

message HealthCheckResponse {
    enum ServingStatus {
        UNKNOWN = 0;
        SERVING = 1;
        NOT_SERVING = 2;
    }
    ServingStatus status = 1;
    optional string reason = 2;
}

///////////////////////////////////////////////////////////////////////////////////

service PosManagement {
    rpc UpdateVoluemMeta (UpdateVoluemMetaRequest) returns (PosResponse) {}
    rpc GetVolumeList (GetVolumeListRequest) returns (VolumeListResponse) {}
}

message UpdateVoluemMetaRequest {
    string array_name = 1;
}

enum PosResult {
    SUCCESS = 0;
    FAIL = 1;
}

message PosResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message GetVolumeListRequest {
    string array_name = 1;
}

message VolumeInfo {
    string volume_name = 1;
    uint64 total_capacity = 2;
    uint64 used_capacity = 3;

    enum VolumeState {
        UNKNOWN = 0;
        MOUNTED = 1;
        UNMOUNTED = 2;
    }
    VolumeState state = 4;
}

message VolumeListResponse {
    PosResult result = 1;
    optional string reason = 2;
    repeated VolumeInfo infos = 3;
}

///////////////////////////////////////////////////////////////////////////////////

service PosIo {
    rpc ReadBlocks(ReadBlocksRequest) returns (ReadBlocksResponse) {}
    rpc WriteBlocks(WriteBlocksRequest) returns (WriteBlocksResponse) {}
    rpc WriteHostBlocks(WriteHostBlocksRequest) returns (WriteHostBlocksResponse) {}
    rpc ReadBlocksSync(ReadBlocksSyncRequest) returns (ReadBlocksSyncResponse) {} // FakePOS-only sync API
    rpc WriteBlocksSync(WriteBlocksSyncRequest) returns (WriteBlocksSyncResponse) {} // FakePOS-only sync API
    rpc WriteHostBlocksSync(WriteHostBlocksSyncRequest) returns (WriteHostBlocksSyncResponse) {} // FakePOS-only sync API
    rpc CompleteHostWrite(CompleteHostWriteRequest) returns (CompleteHostWriteResponse) {}
}

message Chunk {
    bytes content = 1;  // 512B data
}

message ReadBlocksRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 rba = 3;
    uint64 num_blocks = 4;
}

message ReadBlocksResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message WriteBlocksRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 rba = 3;
    uint64 num_blocks = 4;
    uint64 lsn = 5;
    repeated Chunk data = 6;
}

message WriteBlocksResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message WriteBlocksSyncRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 rba = 3;
    uint64 num_blocks = 4;
    repeated Chunk data = 5;
}

message WriteBlocksSyncResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message ReadBlocksSyncRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 rba = 3;
    uint64 num_blocks = 4;
}

message ReadBlocksSyncResponse {
    PosResult result = 1;
    optional string reason = 2;
    repeated Chunk data = 3;
}

message WriteHostBlocksRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 lsn = 3;
}

message WriteHostBlocksResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message WriteHostBlocksSyncRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 lsn = 3;
}

message WriteHostBlocksSyncResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message CompleteHostWriteRequest {
    string array_name = 1;
    string volume_name = 2;
    uint64 lsn = 3;
}

message CompleteHostWriteResponse {
    PosResult result = 1;
    optional string reason = 2;
}

///////////////////////////////////////////////////////////////////////////////////

service ReplicationControl {
    rpc StartVolumeSync(StartVolumeSyncRequest) returns (StartVolumeSyncResponse) {}
    rpc FinishVolumeSync(FinishVolumeSyncRequest) returns (FinishVolumeSyncResponse) {}
    rpc SuspendWrite(SuspendWriteRequest) returns (SuspendWriteResponse) {}
    rpc ResumeWrite(ResumeWriteRequest) returns (ResumeWriteResponse) {}
    rpc ChangeReplicationRole(ChangeReplicationRoleRequest) returns (ChangeReplicationRoleResponse) {}
}

message StartVolumeSyncRequest {
    string array_name = 1;
    string volume_name = 2;
    bool is_primary = 3;
}

message StartVolumeSyncResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message FinishVolumeSyncRequest {
    string array_name = 1;
    string volume_name = 2;
}

message FinishVolumeSyncResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message SuspendWriteRequest {
    string array_name = 1;
    string volume_name = 2;
}

message SuspendWriteResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message ResumeWriteRequest {
    string array_name = 1;
    string volume_name = 2;
}

message ResumeWriteResponse {
    PosResult result = 1;
    optional string reason = 2;
}

message ChangeReplicationRoleRequest {
    string array_name = 1;
    string volume_name = 2;
    string role = 3;
}

message ChangeReplicationRoleResponse {
    PosResult result = 1;
    optional string reason = 2;
}